File: output_20260320_065752/claude-opus/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@ predicate person(struct person *p; struct person *s) =
      p->name |-> _ &*& p->spouse |-> s; @*/

struct person *create_person()
  //@ requires true;
  //@ ensures person(result, 0);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close person(p, 0);
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires person(this, ?s1) &*& person(other, ?s2);
  //@ ensures person(this, other) &*& person(other, this);
{
  //@ open person(this, s1);
  //@ open person(other, s2);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this, other);
  //@ close person(other, this);
}

struct person* person_get_spouse(struct person* this)
  //@ requires person(this, ?s);
  //@ ensures person(this, s) &*& result == s;
{
  //@ open person(this, s);
  struct person* result = this->spouse;
  //@ close person(this, s);
  return result;
}

void divorce(struct person* this)
  //@ requires person(this, ?s) &*& s != 0 &*& person(s, this);
  //@ ensures person(this, 0) &*& person(s, 0);
{
  //@ open person(this, s);
  //@ open person(s, this);
  s->spouse = 0;
  this->spouse = 0;
  //@ close person(this, 0);
  //@ close person(s, 0);
}

void die(struct person *this)
  //@ requires person(this, ?s);
  //@ ensures true;
{
  //@ open person(this, s);
  if(this->spouse != 0) {
    struct person* sp = this->spouse;
    //@ open person(sp, this);
    sp->spouse = 0;
    //@ close person(sp, 0);
  }
  free(this);
}

int main() 
  //@ requires true;
  //@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260320_065752/qwen3/java/Linklist/Spouse.java
/*@ predicate person(Person p; Person spouse) = p.spouse |-> spouse; @*/

class Person {

    protected Person spouse;

    
    //@ requires person(this, _) &*& person(other, _);
    //@ ensures person(this, other) &*& person(other, this);
    public  void spouse_symm()
    {
        //@ open person(this, _);
        //@ open person(other, _);
        spouse = other;
        other.spouse = this;
        //@ close person(this, other);
        //@ close person(other, this);
    }

    //@ ensures person(this, null);
    public Person()
    {
        //@ close person(this, null);
    }
    
    //@ requires person(this, ?s);
    //@ ensures person(this, s) &*& result == s;
    public Person getSpouse()
    {
        //@ open person(this, _);
        //@ close person(this, spouse);
        return spouse;
    }
    
    //@ requires person(this, _) &*& person(other, _);
    //@ ensures person(this, other) &*& person(other, this);
    protected void setSpouse(Person other)
    {
        //@ open person(this, _);
        //@ open person(other, _);
        spouse = other;
        other.spouse = this;
        //@ close person(this, other);
        //@ close person(other, this);
    }
    
    //@ requires person(this, ?s) &*& s != null &*& person(s, this);
    //@ ensures person(this, null) &*& person(s, null);
    protected void clearSpouse()
    {
        //@ open person(this, ?s);
        //@ open person(s, this);
        spouse.spouse = null;
        spouse = null;
        //@ close person(this, null);
        //@ close person(s, null);
    }
    
    //@ requires person(this, _) &*& person(other, _);
    //@ ensures person(this, other) &*& person(other, this);
    void marry(Person other)
    {
        other.setSpouse(this);
    }
    
    //@ requires person(this, ?s) &*& s != null &*& person(s, this);
    //@ ensures person(this, null) &*& person(s, null);
    void divorce()
    {
        spouse.clearSpouse();
    }

}

class Program {

    //@ requires true;
    //@ ensures true;
    public static void main(String[] args)
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}